internal/gardenclient/shoot_client.go

```diff
@@ -0,0 +1,273 @@
+type shootKubeconfigRequest struct {
+	// cluster holds all the cluster on which the kube-apiserver can be reached
+	clusters []cluster
+	// namespace is the namespace where the shoot resides
+	namespace string
+	// shootName is the name of the shoot
+	shootName string
+	// gardenClusterIdentity is the cluster identifier of the garden cluster.
+	gardenClusterIdentity string
+}
+
+// cluster holds the data to describe and connect to a kubernetes cluster
+type cluster struct {
+	// name is the name of the shoot advertised address, usually "external", "internal" or "unmanaged"
+	name string
+	// apiServerHost is the host of the kube-apiserver
+	apiServerHost string
+
+	// caCert holds the ca certificate for the cluster
+	//+optional
+	caCert []byte
+}
+
+// ExecPluginConfig contains a reference to the garden and shoot cluster
+type ExecPluginConfig struct {
+	// ShootRef references the shoot cluster
+	ShootRef ShootRef `json:"shootRef"`
+	// GardenClusterIdentity is the cluster identifier of the garden cluster.
+	// See cluster-identity ConfigMap in kube-system namespace of the garden cluster
+	GardenClusterIdentity string `json:"gardenClusterIdentity"`
+}
+
+// ShootRef references the shoot cluster by namespace and name
+type ShootRef struct {
+	// Namespace is the namespace of the shoot cluster
+	Namespace string `json:"namespace"`
+	// Name is the name of the shoot cluster
+	Name string `json:"name"`
+}
+
+// validate validates the kubeconfig request by ensuring that all required fields are set
+func (k *shootKubeconfigRequest) validate() error {
+	if len(k.clusters) == 0 {
+		return errors.New("missing clusters")
+	}
+
+	for n, cluster := range k.clusters {
+		if cluster.name == "" {
+			return fmt.Errorf("no name defined for cluster[%d]", n)
+		}
+
+		if cluster.apiServerHost == "" {
+			return fmt.Errorf("no api server host defined for cluster[%d]", n)
+		}
+	}
+
+	if k.namespace == "" {
+		return errors.New("no namespace defined for kubeconfig request")
+	}
+
+	if k.shootName == "" {
+		return errors.New("no shoot name defined for kubeconfig request")
+	}
+
+	if k.gardenClusterIdentity == "" {
+		return errors.New("no garden cluster identity defined for kubeconfig request")
+	}
+
+	return nil
+}
+
+// generate generates a Kubernetes kubeconfig for communicating with the kube-apiserver
+// by exec'ing the gardenlogin plugin, which fetches a client certificate.
+// If legacy is false, the shoot reference and garden cluster identity is passed via the cluster extensions,
+// which is supported starting with kubectl version v1.20.0.
+// If legacy is true, the shoot reference and garden cluster identity are passed as command line flags to the plugin
+func (k *shootKubeconfigRequest) generate(legacy bool) ([]byte, error) {
+	authName := fmt.Sprintf("%s--%s", k.namespace, k.shootName) // TODO instead of namespace, use project? But this would require an additional call
+	name := fmt.Sprintf("%s-%s", authName, k.clusters[0].name)
```

Comment:
Are we sure that the first one is always the external endpoint?

Reply:
no, but there is currently no better indicator, except maybe for hardcoding the name but even then it's not guaranteed that there is such an entry
```go
// ToAdvertisedAddresses returns list of advertised addresses on a Shoot cluster.
func (o *Operation) ToAdvertisedAddresses() []gardencorev1beta1.ShootAdvertisedAddress {
	var addresses []gardencorev1beta1.ShootAdvertisedAddress

	if o.Shoot == nil {
		return addresses
	}

	if o.Shoot.ExternalClusterDomain != nil && len(*o.Shoot.ExternalClusterDomain) > 0 {
		addresses = append(addresses, gardencorev1beta1.ShootAdvertisedAddress{
			Name: "external",
			URL:  "https://" + gutil.GetAPIServerDomain(*o.Shoot.ExternalClusterDomain),
		})
	}

	if len(o.Shoot.InternalClusterDomain) > 0 {
		addresses = append(addresses, gardencorev1beta1.ShootAdvertisedAddress{
			Name: "internal",
			URL:  "https://" + gutil.GetAPIServerDomain(o.Shoot.InternalClusterDomain),
		})
	}

	if len(o.APIServerAddress) > 0 && len(addresses) == 0 {
		addresses = append(addresses, gardencorev1beta1.ShootAdvertisedAddress{
			Name: "unmanaged",
			URL:  "https://" + o.APIServerAddress,
		})
	}

	return addresses
}
```
that's why I suggest we just take the first one as "preferred" / current context